tfx/experimental/pipeline_testing/stub_launcher_test.py

```diff
@@ -0,0 +1,188 @@
+from tfx.orchestration import data_types
+from tfx.orchestration import metadata
+from tfx.orchestration import publisher
+from tfx.orchestration.launcher import test_utils
+from tfx.types import channel_utils
+from tfx.utils import io_utils
+
+class CustomStubExecutor(base_stub_executor.BaseStubExecutor):
+  """Example of custom stub executor."""
+  def Do(self, input_dict: Dict[Text, List[types.Artifact]],
+         output_dict: Dict[Text, List[types.Artifact]],
+         exec_properties: Dict[Text, Any]) -> None:
+    super(CustomStubExecutor, self).Do(input_dict,
+                                       output_dict,
+                                       exec_properties)
+    absl.logging.info("Running CustomStubExecutor")
+    for artifact_list in output_dict.values():
+      for artifact in artifact_list:
+        custom_output_path = os.path.join(artifact.uri, "result.txt")
+        io_utils.write_string_file(custom_output_path, "custom component")
+
+class StubLauncherTest(tf.test.TestCase):
+
+  def setUp(self):
+    super(StubLauncherTest, self).setUp()
+    test_dir = os.path.join(
+        os.environ.get('TEST_UNDECLARED_OUTPUTS_DIR', self.get_temp_dir()),
+        self._testMethodName)
+
+    connection_config = metadata_store_pb2.ConnectionConfig()
+    connection_config.sqlite.SetInParent()
+    self.metadata_connection = metadata.Metadata(connection_config)
+
+    self.pipeline_root = os.path.join(test_dir, 'Test')
+    self.input_dir = os.path.join(test_dir, 'input')
+    self.output_dir = os.path.join(test_dir, 'output')
+    self.record_dir = os.path.join(test_dir, 'record')
+    tf.io.gfile.makedirs(self.input_dir)
+    tf.io.gfile.makedirs(self.output_dir)
+    tf.io.gfile.makedirs(self.record_dir)
+
+    input_artifact = test_utils._InputArtifact()  # pylint: disable=protected-access
+    input_artifact.uri = os.path.join(self.input_dir, "result.txt")
+    output_artifact = test_utils._OutputArtifact()  # pylint: disable=protected-access
+    output_artifact.uri = os.path.join(self.output_dir, "result.txt")
+    self.component = test_utils._FakeComponent(  # pylint: disable=protected-access
+        name='FakeComponent',
+        input_channel=channel_utils.as_channel([input_artifact]),
+        output_channel=channel_utils.as_channel([output_artifact]))
+    self.driver_args = data_types.DriverArgs(enable_cache=True)
+
+    self.pipeline_info = data_types.PipelineInfo(
+        pipeline_name='Test', pipeline_root=self.pipeline_root, run_id='123')
+
+  @mock.patch.object(publisher, 'Publisher')
+  def testCustomStubExecutor(self, mock_publisher):
+    # verify whether custom stub executor substitution works
+    mock_publisher.return_value.publish_execution.return_value = {}
+
+    record_file = os.path.join(self.record_dir, 'output', 'recorded.txt')
+    io_utils.write_string_file(record_file, "hello world")
+    component_map = \
+        {'_FakeComponent.FakeComponent': CustomStubExecutor}
+
+    MyStubLauncher = \
```

Comment:
Variable name "MyStubLauncher" doesn't conform to u'^[a-z][a-z0-9_]*$' pattern [invalid-name]. Maybe `my_stub_launcher`?